Dz8

angryQ opened the pull request.

angryQ committed:
ДЗ-8 и дополнительные ДЗ по модулю 7

Были проблемы с нотником, поэтому пока он был на ремонте, работала за чужим когда была возможность. Поэтому нет промежуточных коммитов(

angryQ committed:
Удалила DZ7 который смерджен с матером

Как он сюда попал не поняла, в структуре проекта его не было, так же как и в списке файлов при коммите.

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/ContentViewController.swift:
```diff
@@ -0,0 +1,69 @@
+        startButton.isHidden = true
+        presentTextLabel.text = presentText
+        
+        if let unwrappedImageView = imageView {
```
для свойства imageView уже сразу используется force unwrap.
force unwrap обычно разрешается использовать для @IBOutlet свойств.
А значит можно использовать везде сразу imageView без дополнительного извлечения.

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/ContentViewController.swift:
```diff
@@ -0,0 +1,69 @@
+        if let mainViewController = mainStoryboard.instantiateViewController(
+            withIdentifier: "TabBar") as? UITabBarController {
+            
+            window = UIWindow(frame: UIScreen.main.bounds)
+            window?.rootViewController = mainViewController
+            window?.makeKeyAndVisible()
+            
+            /*Как правильно выходить из PаgeViewController'а? При использовании кода ниже,
+             PageViewController продолжает отображаться на экране, просто скрыт раснятуным на весь экран
+             модальным окном, и это не очень хорошо. При попытке удалить его из памяти, также и удаляется
+             модальное окно с TabBarController'ом (mainViewController). В итоге я создала еще один UIWindow,
+             и как я поняла UIApplication может содержать несколько UIWindow, а тогда как система поняла что нужно
+             отобразить новое окно и что стало с предыдущим?
+             
+             mainViewController.modalPresentationStyle = .fullScreen
+             present(mainViewController, animated: true, completion: nil)
+             */
```
В данном случае самым оптимальным решением будет поменять rootViewController для текущего экрана.
У любой UIView есть свойство window, на которой отображается данная вью.

`self.view.window?.rootViewController = mainViewController`

Если необходимо с анимацией, то можно так:
```
@IBAction func startWatchHomeWork() {
        
        let mainStoryboard = UIStoryboard (name: "Main", bundle: nil)
        
        if let mainViewController = mainStoryboard.instantiateViewController(
            withIdentifier: "TabBar") as? UITabBarController {
            
            guard let window = view.window else { return }
            
            window.rootViewController = mainViewController
            let options: UIView.AnimationOptions = .transitionCrossDissolve
            let duration: TimeInterval = 2
            UIView.transition(with: window, duration: duration, options: options, animations: nil)
        }
    }
```

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/ContentViewController.swift:
```diff
@@ -0,0 +1,69 @@
+        }
+        pageControl.numberOfPages = numberOfPages
+        pageControl.currentPage = currentPage
+        pageControl.isHidden = pageControlIsHidden
+        startButton.isHidden = startButtonIsHidden
```
Эти два свойства можно вычислять в зависимости от numberOfPages и currentPage.
Попробуете найти закономерность самостоятельно?

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/PageViewController.swift:
```diff
@@ -0,0 +1,78 @@
+//Класс, который управляет другими вью контроллерами, которые будут отображать контент моей презентации
+class PageViewController: UIPageViewController {
+    
+    let presentScreenContent = ["На первой табе находится первое д/з с галереей",
+                                "На второй табе находится д/з с отображением четырех изображений на одном экране",
+                                "На третьей табе д/з с использованием UISegmentedControl",
+                                "На четвертой табе реализация Container View Controller (дополнительное д/з за 7 модуль)",
+                                "А вся эта презентация демонстрирует применение UIPageViewController (дополнительное д/з за 7 модуль)"
+        
+    ]
+    let imageNameArray = ["pic-2","pic-5","pic-9","pic-6","autumn"]
+    let buttonIsHidden = [true, true, true, true, false]
+    let pageControlIsHidden = [false, false, false, false, true]
```
Реализуете модель данных с помощью структуры?

Почему использование структуры будет безопаснее, эффективнее, нагляднее, проще?

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/PageViewController.swift:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+    
+    //Метод который позволяет нам создать view controller
+    func showViewControllerAtIndex(_ index: Int) -> ContentViewController? {
```
show - показать
make, create - создать.

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/PageViewController.swift:
```diff
@@ -0,0 +1,78 @@
+    //Метод который позволяет нам создать view controller
+    func showViewControllerAtIndex(_ index: Int) -> ContentViewController? {
+        
+        guard index >= 0 else { return nil }
+        
+        guard index < presentScreenContent.count else { return nil }
```
можно ли объединить с помощью &&?

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/PageViewController.swift:
```diff
@@ -0,0 +1,78 @@
+
+import UIKit
+
+//Класс, который управляет другими вью контроллерами, которые будут отображать контент моей презентации
+class PageViewController: UIPageViewController {
```
Реализация onboarding очень тронула меня😍😍😍

Полноценный пользовательский опыт! Просто шикарно🔥🔥🔥🔥🔥

ios-review commented on DZ8/DZ8/FourImagesOnScreen/FourImagesViewController.swift:
```diff
@@ -0,0 +1,94 @@
+        
+        self.view.backgroundColor = Colors.backgroundColor
+        
+        //Устанавливаю размер и положение вью, вычитаю из ширины по 20 точек слева и справа, и от высоты нижний
+        //и верхний отступ и приблизительную высоту таббара
```
Комментарии к коду топ👍

ios-review commented on DZ8/DZ8/FourImagesOnScreen/FourImagesViewController.swift:
```diff
@@ -0,0 +1,94 @@
+        
+        //Устанавливаю размер и положение вью, вычитаю из ширины по 20 точек слева и справа, и от высоты нижний
+        //и верхний отступ и приблизительную высоту таббара
+        holderView.frame = CGRect(x: 20, y: 20, width: self.view.frame.width - 40, height: self.view.frame.height - 80)
```
В методе viewDidLoad еще неизвестен точно размер и положение вью у контроллера.
self.view.frame в данном методе установлен значением, которое указано в сториборде.
А в сториборде это значение указано конкретное для определенного размера.

В методе viewWillLayoutSubviews() bounds для view уже известен.

Про жизненный цикл контроллера можно прочитать по ссылке https://medium.com/@felicity.johnson.mail/viewcontroller-lifestyle-e304e1078bc5

ios-review commented on DZ8/DZ8/FourImagesOnScreen/FourImagesViewController.swift:
```diff
@@ -0,0 +1,94 @@
+        var startIndexInSubArray = 0 //начальный индекс в срезе элементов массива для прохождения по нему циклом
+        var endIndexInSubArray = 1 //конечный индекс в срезе элементов массива для прохождения по нему циклом
+        
+        //Не разобралась как вычислить высоту таббара без его аутлета, который не могу добавить в этот вью контроллер,
+        //поэтому использовала приблизительное значение
+        let tabBarHeight: CGFloat = 40
```
`self.tabBarController?.tabBar.frame`

ios-review commented on DZ8/DZ8/FourImagesOnScreen/FourImagesViewController.swift:
```diff
@@ -0,0 +1,94 @@
+        var xPos: CGFloat = 0 // "х" для точки отсчета позиции элемента
+        
+        
+        /*Высота и ширина картинки. С высотой какая-то проблема, так и не поняла откуда берутся лишние точки.
+        Например, получила высоту вью holderView.bounds.height, вижу ее полностью на экране, делю ее высоту напополам
+         и получаю высоту одной строки. Из получившейся высоты строки вычитаю высоту лэйбла и отступы под картинкой
+         и лэйблом, оставшаяся часть от высоты строки идет на высоту картинки, но почему-то картинки не помещаются
+         во вью. Не поняла что я считаю не так и что не учитываю. В итоге пришлось вычитать из heightPerItem
+         еще 50 чтобы все поместилось */
+        let widthPerItem = holderView.bounds.width / countOfRowsOnScreen - verticalPadding
+        let heightPerItem = holderViewHeightWithoutTabbar / countOfRowsOnScreen - paddingUnderImage - labelHeight
+            - paddingUnderLabel - 50
```
Расчет ведется в методе viewDidLoad, в котором истинный размер holderView еще не может быть точно установлен.

Возможно проблема в этом.
Попробуете проверить, реализовав расчет размеров в viewWillLayoutSubviews()?

ios-review commented on DZ8/DZ8/FourImagesOnScreen/FourImagesViewController.swift:
```diff
@@ -0,0 +1,94 @@
+            - paddingUnderLabel - 50
+        
+        //Проходим по всему массиву элементов, беря по 2 элемента и формируя из них строку из двух элементов
+        while endIndexInSubArray < store.images.count {
+            
+            for item in store.images[startIndexInSubArray...endIndexInSubArray] {
+                
+                //Создаем картинку и лэйбл, устанавливаем в них значения и добавляем их на вью
+                let imageView = UIImageView(image: item.image)
+                
+                imageView.frame = CGRect(x: xPos, y: yPos, width: widthPerItem, height: heightPerItem)
+                imageView.imgViewCorners()
+                
+                let label = UILabel(frame: CGRect(x: xPos, y: yPos + heightPerItem + paddingUnderImage,
+                                                  width: widthPerItem, height: labelHeight))
+                label.text = item.name
+                
+                holderView.addSubview(imageView)
+                holderView.addSubview(label)
+                
+                //устанавливаем "х" для второго элемента во вторую колонку
+                xPos = holderView.frame.width / countOfRowsOnScreen + verticalPadding
+            }
+            //Устанавливаем "х" в первую колонку, а "у" на вторую строку
+            yPos = holderViewHeightWithoutTabbar / countOfRowsOnScreen
+            xPos = 0
+            
+            startIndexInSubArray += 2
+            endIndexInSubArray += 2
```
Крутой алгоритм😎

ios-review commented on DZ8/DZ8/FourImagesOnScreen/ImageCollection.swift:
```diff
@@ -0,0 +1,35 @@
+    var image: UIImage
+}
+
+class ImageCollection {
+    
+    var images = [Item]()
+    
+    init() {
+        setup()
+    }
+    
+    func setup() {
+        
+        let item1 = Item(name: "Картинка #1", image: UIImage(named: "pic-10")!)
+        let item2 = Item(name: "Картинка #2", image: UIImage(named: "pic-3")!)
+        let item3 = Item(name: "Картинка #3", image: UIImage(named: "pic-9")!)
+        let item4 = Item(name: "Картинка #4", image: UIImage(named: "pic-8")!)
+        
+        self.images = [item1, item2, item3, item4]
+    }
+}
```
Можно немного проще:
```
class ImageCollection {
 
    let images = [Item(name: "Картинка #1", image: UIImage(named: "pic-10")!),
                  Item(name: "Картинка #2", image: UIImage(named: "pic-3")!),
                  Item(name: "Картинка #3", image: UIImage(named: "pic-9")!),
                  Item(name: "Картинка #4", image: UIImage(named: "pic-8")!)]
 
  }
```

ios-review replied:
Старайтесь использовать наиболее короткий синтаксис.

ios-review commented on DZ8/DZ8/Gallery/GalleryViewController.swift:
```diff
@@ -0,0 +1,53 @@
+        
+        view.backgroundColor = Colors.backgroundColor
+        
+        if let unwrappedImageView = galleryImageView {
```
можно не использовать извлечение значения, так как galleryImageView уже под force unwrap, что допустимо для @IBOutlet свойств.

ios-review commented on DZ8/DZ8/Gallery/GalleryViewController.swift:
```diff
@@ -0,0 +1,53 @@
+    //Пробовала анимацию изображений
+    private func animateImage(imageName: String) {
+        
+        UIView.transition(with: galleryImageView, duration: 0.1, options: .transitionCrossDissolve, animations: {
```
👌

ios-review commented on DZ8/DZ8/SegmentedControl/SegmentedViewController.swift:
```diff
@@ -0,0 +1,93 @@
+    private func addConstraintsToSubview(subview: UIView) {
+        
+        subview.translatesAutoresizingMaskIntoConstraints = false
+        subview.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 0).isActive = true
+        subview.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: 0).isActive = true
+        subview.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 0).isActive = true
+        subview.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: 0).isActive = true
```
Для массива констрейнтов рекомендуется делать так:
`NSLayoutConstraint.activate([/*здесь массив констрейнтов*/])`

ios-review commented on DZ8/DZ8/SegmentedControl/SegmentedViewController.swift:
```diff
@@ -0,0 +1,93 @@
+        subview.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: 0).isActive = true
+    }
+    
+    override func didReceiveMemoryWarning() {
+        
+        //Нужно ли это делать??
+        blueViewController.view = nil
+        greenViewController.view = nil
+        purpleViewController.view = nil
+    }
```
Данные ресурсы контролирует UIKit, поэтому в случае нехватки памяти фреймворк это делает за нас.

В этом методе рекомендуется очищать те ресурсы, которые контролируем только мы и которые можно будет восстановить.

ios-review commented on DZ8/DZ8/Utilities/Constants.swift:
```diff
@@ -0,0 +1,108 @@
+    }
+    func sizeButton() {
+        //titleEdgeInsets = UIEdgeInsets(top: 0, left: 5, bottom: 0, right: 5)
+        translatesAutoresizingMaskIntoConstraints = false
+        widthAnchor.constraint(equalToConstant: (UIScreen.main.bounds.width - 60) / 3).isActive = true
+        heightAnchor.constraint(equalToConstant: 40).isActive = true
```
Обычно определение размеров вью, включая установку констрейнтов для ширины/высоты, является ответственностью клиента данной вью.

Использовать UIScreen в классах UIView в общем случае не рекомендуется.
Как и везде.
Такая зависимость может привести к очень интересным багам.
Попробуете продумать хоть один такой?

Плюс клиент может установить свои значения ширины и высоты, которые будут конфликтовать с данными.

ios-review commented on DZ8/DZ8/Utilities/Constants.swift:
```diff
@@ -0,0 +1,108 @@
+    static let medium = "AvenirNext-Medium"
+    static let regular = "AvenirNext-Regular"
+}
+class CustomButton: UIButton {
```
Подкласс UIButton расположен в файле Constants

ios-review commented on DZ8/DZ8/Support/AppDelegate.swift:
```diff
@@ -0,0 +1,40 @@
+        
+        window = UIWindow(frame: UIScreen.main.bounds)
+        
+        let launchedBefore = UserDefaults.standard.bool(forKey: "hasLaunched")
+        let launchStoryboard = UIStoryboard(name: "Onboarding", bundle: nil)
+        let mainStoryboard = UIStoryboard (name: "Main", bundle: nil)
+        var vc: UIViewController
+        
+        if launchedBefore {
+            vc = mainStoryboard.instantiateInitialViewController()!
+        } else {
+            vc = launchStoryboard.instantiateViewController(withIdentifier: "PageViewController")
+        }
+        
+        UserDefaults.standard.set(true, forKey: "hasLaunched")
```
🤩
Порадовало, что onboarding отображается только один раз.

ios-review commented on DZ8/DZ8/SegmentedControl/SegmentedViewController.swift:
```diff
@@ -0,0 +1,93 @@
+    
+    @IBAction func segmentedControlIndexChanged(segment: UISegmentedControl) {
+        
+        switch segmentedControl.selectedSegmentIndex
+        {
+        case 0:
+            remove(childController: purpleViewController)
+            remove(childController: blueViewController)
+            add(childController: greenViewController)
+        case 1:
+            remove(childController: greenViewController)
+            remove(childController: purpleViewController)
+            add(childController: blueViewController)
+        case 2:
+            remove(childController: greenViewController)
+            remove(childController: blueViewController)
+            add(childController: purpleViewController)
+        default:
+            break
+        }
```
```
for (index, vc) in [pVC, gVC, bVC].enumerated() {
    if index == segmentedControl.selectedSegmentIndex {
        add(vc)
    } else {
        remove(vc)
    }
}
```

ios-review commented on DZ8/DZ8/AdditionalTask-Views/AdditionalTaskViewController.swift:
```diff
@@ -0,0 +1,126 @@
+
+import UIKit
+
+class AdditionalTaskViewController: UIViewController {
```
Классная реализация!

Предлагаю немного модифицировать зависимости контроллеров.
Все контент вью контроллеры должны передаваться в ContainerViewController, но не более 6. При передаче больше 6 можно кинуть assert.

ContainerViewController может иметь следующий интерфейс:
```
class ContainerViewController: UIViewController {
    
    var buttonsStackView: UIStackView
    var childVCStackView: UIStackView
    
    private var childs: [UIViewController] = []
    private var defaultVC: UIViewController?
    
    func addVC(_ vc: UIViewController, buttonTitle: String) {
        assert(childs.count < 6, "Too many child ViewControllers: only 6 allowed")
        // Сохраняем контроллер, создаем кнопку, показываем кнопку.
    }
    
    func setPlaceholder(_ vc: UIViewController) {
        // Сохраняем контроллер, который показывается в случае, если ни один контент контроллер не показан
    }
    
   @objc func showHideContentVC(_ sender: UIButton) {
        // Показываем или скрываем контент контроллер, который соответствует кнопке
        // Если все контент контроллеры скрыты, то показываем placeholder
    }

    private func addChildVC(childVC: UIViewController) {
        // Функция для добавления контроллера в иерархию и его показа
    }
    
    private func removeChildVC(childVC: UIViewController) {
        // Функция для удаления контроллера из иерархии и его скрытия
    }
}

```

Этот контроллер должен быть создан из кода, сконфигурирован с помощью других контент контоллеров извне, назначен как главный контроллер всего приложения.
Как создать UI из кода, не используя сториборд, описано по ссылке https://stackoverflow.com/questions/57963647/how-to-create-ui-programmatically-in-xcode-11
Смотреть нужно правильный ответ, который помечен зеленой галочкой.

В SceneDelegate, где должен создаваться ContainerViewController может быть следующий код для создания UI:
```
        let conVC = ContainerViewController()
        conVC.addVC(UIViewController(), buttonTitle: "one")
        conVC.addVC(UIViewController(), buttonTitle: "two")
        conVC.addVC(UIViewController(), buttonTitle: "three")
        conVC.addVC(UIViewController(), buttonTitle: "four")
        conVC.addVC(UIViewController(), buttonTitle: "five")
        conVC.addVC(UIViewController(), buttonTitle: "six")
        
        let defaultVC = UIViewController()
        defaultVC.view.backgroundColor = .lightGray
        conVC.setDefaultPlaceholder(defaultVC)
        
        window?.rootViewController = conVC
        window?.makeKeyAndVisible()
```

ios-review reviewed (changes requested):
Готов вечно признаваться в любви к вашему коду!
Шикарные реализации!
Супер-молодец👍

Поработаете еще над альтернативами?

angryQ committed:
Первая часть правок по замечаниям Дмитрия

Добавила правки по замечаниям Дмитрия, кроме задачи с изменением зависимости контроллеров, ее начну делать после этого коммита, чтобы было к чему вернуться если я все вдруг поломаю )

angryQ committed:
Переход из onboarding презентации на таббар из SceneDelegate

Кое как разобралась как сделать переход из onboarding презентации на таббар, создаваемый в SceneDelegate. Осталось разобраться почему не обрабатываются события на нажатия кнопок в дополнительном задании.

angryQ committed:
Починила работу кнопок на табе 4, удалила лишние файлы

Удалила папку  AdditionalTask-Views, раньше она содержала файлы для табы 4, теперь таба 4 создается в файле ContainerViewController

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/ContentViewController.swift:
```diff
@@ -31,39 +31,21 @@ class ContentViewController: UIViewController {
         pageControl.currentPage = currentPage
-        pageControl.isHidden = pageControlIsHidden
-        startButton.isHidden = startButtonIsHidden
+        if currentPage == numberOfPages - 1 {
+            pageControl.isHidden = true
+            startButton.isHidden = false
+        }
```
👍

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/PageViewController.swift:
```diff
@@ -11,46 +11,55 @@ import UIKit
+        }
+    }
+    
+    var pageArray: [OnboardingPage] = []
```
```
let pageArray = [OnboardingPage(content: "На первой табе находится первое д/з с галереей", imageName: "pic-2"),
                 OnboardingPage(content: "На второй табе находится д/з с отображением четырех изображений на одном экране",
                                  imageName: "pic-5")]
```

ios-review commented on DZ8/DZ8/AdditionalTask-PageViewController/PageViewController.swift:
```diff
@@ -11,46 +11,55 @@ import UIKit
         super.viewDidLoad()
         
         dataSource = self
+        pageArray = [pageOne, pageTwo, pageThree, pageFour, pageFive]
```
все, что можно проинициализировать по месту объявления, инициализировать лучше по месту объявления.

ios-review commented on DZ8/DZ8/Main/ContainerViewController.swift:
```diff
@@ -0,0 +1,150 @@
+        
+        childs.append(vc)
+        addChildVC(childVC: vc)
+        holderStackView.addArrangedSubview(childVCStackView)
```
Эта строчка должна быть в viewDidLoad.
Достаточно за все время жизни контроллера один раз добавить childVCStackView в holderStackView

ios-review commented on DZ8/DZ8/Main/ContainerViewController.swift:
```diff
@@ -0,0 +1,150 @@
+    override func viewWillLayoutSubviews() {
+        super.viewWillLayoutSubviews()
+        
+        configureHolderStackView()
+        configureViewsStackView()
```
Эти строки тоже могут быть в viewDidLoad.
Установление констрейнтов может быть в любом месте кода, так как констрейнты - это всего лишь правила, на основе которых будет вычислен frame вью.
frame вью может быть установлен вручную или же посчитан с помощью autolayout.
Если устанавливать его вручную, то тогда только в соответствующих методах жизненного цикла.
Если использовать правила autolayout, то frame будет автоматически посчитан на базе данных правил.
При чем этот расчет будет выполнен также в методах жизненного цикла.
Просто это реализовано в UIView классе.
Более подробно про autolayout вы узнаете из модуля 9.

ios-review commented on DZ8/DZ8/Main/ContainerViewController.swift:
```diff
@@ -0,0 +1,150 @@
+        let button = createAButton(buttonTitle: buttonTitle)
+        addButtonToStackView(button)
+        
+        holderStackView.insertArrangedSubview(buttonsStackView, at: 0)
+        configureButtonsStackView()
```
Данный код также может быть вызван всего один раз за все время жизни контроллера.

ios-review commented on DZ8/DZ8/Main/ContainerViewController.swift:
```diff
@@ -0,0 +1,150 @@
+        configureButtonsStackView()
+    }
+    
+    func addButtonToStackView(_ button: UIButton) {
+        buttonsStackView.addArrangedSubview(button)
+    }
```
Кажется, что функция из одной строчки ничего нового не прибавляет, а значит лишняя.

Везде в коде можно использовать просто:
`buttonsStackView.addArrangedSubview(button)`
будет тоже самое.

ios-review commented on DZ8/DZ8/Main/ContainerViewController.swift:
```diff
@@ -0,0 +1,150 @@
+        
+        let index = buttonsStackView.arrangedSubviews.firstIndex(of: sender)!
+        
+        if sender.title(for: .normal) == "Включить" {
```
Заголовок кнопки является контентом, который может быть в любое время изменен другим программистом по заданию контент-менеджера.
Например, на более понятное слово после анализа статистических данных понимания слов.

Логика приложения может зависеть от контента, но только в случаях когда другого выхода нет.

В данном случае можно сделать зависимость на свойство кнопки isSelected https://developer.apple.com/documentation/uikit/uicontrol/1618203-isselected
При нажатию на кнопку менять это значение:
```
sender.isSelected.toogle()
if sender.isSelected {
    show()
} else {
    remove()
}
```

Также можно указать соответствующие заголовки и цвета для данного состояния кнопки при ее создании с помощью https://developer.apple.com/documentation/uikit/uibutton/1624018-settitle и https://developer.apple.com/documentation/uikit/uibutton/1623993-settitlecolor
```
button.setTitle("On", for: .normal)
button.setTitle("Off", for: .selected)
```
или
```
button.setTitleColor(.green, for: .normal)
button.setTitle(.red, for: .selected)
```
После изменения isSelected свойства кнопка автоматически применить нужный заголовок или цвет заголовка в зависимости от состояния.

Заголовок кнопки передается в контроллер через публичный метод.
Указывать заголовок для выключенного состояния внутри контроллера нарушение ответственности.
Что если пользователь класса захочет установить заголовок 1?
Или любой другой, отличный от Вкл/Выкл?

Есть несколько решений:
- передавать через интерфейс заголовок для selected и normal состояний
- передавать заголовок только для normal состояний, но устанавливать внутри контейнер контроллера разные цвета заголовков для selected и normal состояний

Можно выбрать любой.

ios-review commented on DZ8/DZ8/Main/ContainerViewController.swift:
```diff
@@ -0,0 +1,150 @@
+    
+    func configureViewsStackView() {
+        
+        if childVCStackView.subviews.count == 0 {
+            guard let vc = defaultVC else {return}
+            addChildVC(childVC: vc)
+        }
```
Не понятно назначение данного кода.

ios-review commented on DZ8/DZ8/SegmentedControl/SegmentedViewController.swift:
```diff
@@ -76,18 +66,12 @@ class SegmentedViewController: UIViewController {
-        blueViewController.view = nil
-        greenViewController.view = nil
-        purpleViewController.view = nil
+        NSLayoutConstraint.activate([topConstraint, bottomConstraint, leadingConstraint, trailingConstraint])
```
Самая короткая запись:
```
        NSLayoutConstraint.activate([
            subview.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            subview.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
            subview.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
            subview.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor)
        ])
```

ios-review reviewed (changes requested):
🔥

Контейнер еще упростим?

angryQ committed:
Оптимизации по рекомендациям Дмитрия

ios-review commented on DZ8/DZ8/Main/ContainerViewController.swift:
```diff
@@ -74,18 +66,16 @@ class ContainerViewController: UIViewController {
+        } else {
             removeChildVC(childVC: childs[index])
-            sender.setTitle("Включить", for: .normal)
         }
```
Это требование будет дополнительным заданием.
Его можно будет выполнить в отдельной ветке после merge пул реквеста.

Текущее поведение: контроллеры появляются сверху вниз в порядке нажатия кнопок.
Например,
1. Нажали кнопку 2
2. Появился контроллер 2.
3. Нажали кнопку 1.
4. Под контроллером 2 появился контроллер 1.

Как изменить алгоритм так, чтобы контроллеры появлялись в порядке их добавления с помощью метода addVC(_:)?

Должно быть так:
1. Нажали кнопку 2
2. Появился контроллер 2.
3. Нажали кнопку 1.
4. Контроллер 1 появился над контроллером 2.

Должен сохраняться такой же порядок сверху вниз, как в массиве childs.

ios-review reviewed (approved):
👌🔥👍
